// vue-zoom-on-hover/dist/index.umd.js
/*!
 * vue-zoom-on-hover v1.0.4
 * (c) Ivan Alexandrov
 * Released under the GPL-3.0 License.
 */
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global = global || self, global.ZoomOnHover = factory());
}(this, (function () { 'use strict';

  //
  //
  //
  //
  //
  //
  //
  //
  //
  //
  //
  //
  //
  //
  var script = {
    props: ["imgNormal", "imgZoom", "scale", "disabled", "altText"],
    data: function data() {
      return {
        scaleFactor: 1,
        resizeCheckInterval: null,
        zoomed: false
      };
    },
    methods: {
      pageOffset: function pageOffset(el) {
        // -> {x: number, y: number}
        // get the left and top offset of a dom block element
        var rect = el.getBoundingClientRect(),
            scrollLeft = window.pageXOffset || document.documentElement.scrollLeft,
            scrollTop = window.pageYOffset || document.documentElement.scrollTop;
        return {
          y: rect.top + scrollTop,
          x: rect.left + scrollLeft
        };
      },
      touchzoom: function touchzoom(event) {
        if (this.disabled) return;
        this.zoomed = event.type === "pointerover";
        this.move(event);
      },
      zoom: function zoom() {
        if (this.disabled) return;
        this.zoomed = true;
      },
      unzoom: function unzoom() {
        if (this.disabled) return;
        this.zoomed = false;
      },
      move: function move(event) {
        var zoom = this.$refs.zoom;
        var normal = this.$refs.normal;
        if (this.disabled || !this.zoomed || !zoom || !normal) return;
        var offset = this.pageOffset(this.$el);
        var relativeX = event.clientX - offset.x + window.pageXOffset;
        var relativeY = event.clientY - offset.y + window.pageYOffset;
        var normalFactorX = relativeX / normal.offsetWidth;
        var normalFactorY = relativeY / normal.offsetHeight;
        var x = normalFactorX * (zoom.offsetWidth * this.scaleFactor - normal.offsetWidth);
        var y = normalFactorY * (zoom.offsetHeight * this.scaleFactor - normal.offsetHeight);
        zoom.style.left = -x + "px";
        zoom.style.top = -y + "px";
      },
      initEventLoaded: function initEventLoaded() {
        // emit the "loaded" event if all images have been loaded
        var promises = [this.$refs.zoom, this.$refs.normal].map(function (image) {
          return new Promise(function (resolve, reject) {
            image.addEventListener("load", resolve);
            image.addEventListener("error", reject);
          });
        });
        var component = this;
        Promise.all(promises).then(function () {
          component.$emit("loaded");
        });
      },
      initEventResized: function initEventResized() {
        var normal = this.$refs.normal;
        var previousWidth = normal.offsetWidth;
        var previousHeight = normal.offsetHeight;
        var component = this;
        this.resizeCheckInterval = setInterval(function () {
          if (previousWidth != normal.offsetWidth || previousHeight != normal.offsetHeight) {
            previousWidth = normal.offsetWidth;
            previousHeight = normal.offsetHeight;
            component.$emit("resized", {
              width: normal.width,
              height: normal.height,
              fullWidth: normal.naturalWidth,
              fullHeight: normal.naturalHeight
            });
          }
        }, 1000);
      }
    },
    mounted: function mounted() {
      if (this.$props.scale) {
        this.scaleFactor = parseInt(this.$props.scale);
        this.$refs.zoom.style.transform = "scale(" + this.scaleFactor + ")";
      }

      this.initEventLoaded();
      this.initEventResized();
    },
    updated: function updated() {
      this.initEventLoaded();
    },
    beforeDestroy: function beforeDestroy() {
      this.resizeCheckInterval && clearInterval(this.resizeCheckInterval);
    }
  };

  function normalizeComponent(template, style, script, scopeId, isFunctionalTemplate, moduleIdentifier
  /* server only */
  , shadowMode, createInjector, createInjectorSSR, createInjectorShadow) {
    if (typeof shadowMode !== 'boolean') {
      createInjectorSSR = createInjector;
      createInjector = shadowMode;
      shadowMode = false;
    } // Vue.extend constructor export interop.


    var options = typeof script === 'function' ? script.options : script; // render functions

    if (template && template.render) {
      options.render = template.render;
      options.staticRenderFns = template.staticRenderFns;
      options._compiled = true; // functional template

      if (isFunctionalTemplate) {
        options.functional = true;
      }
    } // scopedId


    if (scopeId) {
      options._scopeId = scopeId;
    }

    var hook;

    if (moduleIdentifier) {
      // server build
      hook = function hook(context) {
        // 2.3 injection
        context = context || // cached call
        this.$vnode && this.$vnode.ssrContext || // stateful
        this.parent && this.parent.$vnode && this.parent.$vnode.ssrContext; // functional
        // 2.2 with runInNewContext: true

        if (!context && typeof __VUE_SSR_CONTEXT__ !== 'undefined') {
          context = __VUE_SSR_CONTEXT__;
        } // inject component styles


        if (style) {
          style.call(this, createInjectorSSR(context));
        } // register component module identifier for async chunk inference


        if (context && context._registeredComponents) {
          context._registeredComponents.add(moduleIdentifier);
        }
      }; // used by ssr in case component is cached and beforeCreate
      // never gets called


      options._ssrRegister = hook;
    } else if (style) {
      hook = shadowMode ? function () {
        style.call(this, createInjectorShadow(this.$root.$options.shadowRoot));
      } : function (context) {
        style.call(this, createInjector(context));
      };
    }

    if (hook) {
      if (options.functional) {
        // register for functional component in vue file
        var originalRender = options.render;

        options.render = function renderWithStyleInjection(h, context) {
          hook.call(context);
          return originalRender(h, context);
        };
      } else {
        // inject component registration as beforeCreate hook
        var existing = options.beforeCreate;
        options.beforeCreate = existing ? [].concat(existing, hook) : [hook];
      }
    }

    return script;
  }

  var normalizeComponent_1 = normalizeComponent;

  var isOldIE = typeof navigator !== 'undefined' && /msie [6-9]\\b/.test(navigator.userAgent.toLowerCase());
  function createInjector(context) {
    return function (id, style) {
      return addStyle(id, style);
    };
  }
  var HEAD;
  var styles = {};

  function addStyle(id, css) {
    var group = isOldIE ? css.media || 'default' : id;
    var style = styles[group] || (styles[group] = {
      ids: new Set(),
      styles: []
    });

    if (!style.ids.has(id)) {
      style.ids.add(id);
      var code = css.source;

      if (css.map) {
        // https://developer.chrome.com/devtools/docs/javascript-debugging
        // this makes source maps inside style tags work properly in Chrome
        code += '\n/*# sourceURL=' + css.map.sources[0] + ' */'; // http://stackoverflow.com/a/26603875

        code += '\n/*# sourceMappingURL=data:application/json;base64,' + btoa(unescape(encodeURIComponent(JSON.stringify(css.map)))) + ' */';
      }

      if (!style.element) {
        style.element = document.createElement('style');
        style.element.type = 'text/css';
        if (css.media) style.element.setAttribute('media', css.media);

        if (HEAD === undefined) {
          HEAD = document.head || document.getElementsByTagName('head')[0];
        }

        HEAD.appendChild(style.element);
      }

      if ('styleSheet' in style.element) {
        style.styles.push(code);
        style.element.styleSheet.cssText = style.styles.filter(Boolean).join('\n');
      } else {
        var index = style.ids.size - 1;
        var textNode = document.createTextNode(code);
        var nodes = style.element.childNodes;
        if (nodes[index]) style.element.removeChild(nodes[index]);
        if (nodes.length) style.element.insertBefore(textNode, nodes[index]);else style.element.appendChild(textNode);
      }
    }
  }

  var browser = createInjector;

  /* script */
  var __vue_script__ = script;
  /* template */

  var __vue_render__ = function __vue_render__() {
    var _vm = this;

    var _h = _vm.$createElement;

    var _c = _vm._self._c || _h;

    return _c('div', {
      staticClass: "zoom-on-hover",
      "class": {
        zoomed: _vm.zoomed
      },
      on: {
        "pointerover": _vm.touchzoom,
        "pointerout": _vm.touchzoom,
        "pointermove": _vm.move,
        "pointerenter": _vm.zoom,
        "pointerleave": _vm.unzoom
      }
    }, [_c('img', {
      ref: "normal",
      staticClass: "normal",
      attrs: {
        "src": _vm.imgNormal,
        "alt": _vm.altText
      }
    }), _vm._v(" "), _c('img', {
      ref: "zoom",
      staticClass: "zoom",
      attrs: {
        "src": _vm.imgZoom || _vm.imgNormal,
        "alt": _vm.altText
      }
    })]);
  };

  var __vue_staticRenderFns__ = [];
  /* style */

  var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
    if (!inject) return;
    inject("data-v-0ff054f9_0", {
      source: ".zoom-on-hover[data-v-0ff054f9]{position:relative;overflow:hidden}.zoom-on-hover .normal[data-v-0ff054f9]{width:100%}.zoom-on-hover .zoom[data-v-0ff054f9]{position:absolute;opacity:0;transform-origin:top left}.zoom-on-hover.zoomed .zoom[data-v-0ff054f9]{opacity:1}.zoom-on-hover.zoomed .normal[data-v-0ff054f9]{opacity:0}",
      map: undefined,
      media: undefined
    });
  };
  /* scoped */


  var __vue_scope_id__ = "data-v-0ff054f9";
  /* module identifier */

  var __vue_module_identifier__ = undefined;
  /* functional template */

  var __vue_is_functional_template__ = false;
  /* style inject SSR */

  /* style inject shadow dom */

  var __vue_component__ = /*#__PURE__*/normalizeComponent_1({
    render: __vue_render__,
    staticRenderFns: __vue_staticRenderFns__
  }, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, browser, undefined, undefined);

  var index = {
    install: function install(Vue, options) {
      Vue.component("zoom-on-hover", __vue_component__);
    }
  };

  return index;

})));
